ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,356 @@
+///
+/// f takes as input an element in the slice and outputs the key by which we sort by
+/// follows partially the function signature of `sort_by_key`, see `https://doc.rust-lang.org/src/alloc/slice.rs.html#305-308`
+///
+/// set desc = true for descending ordering
+///
+/// This version of quicksort is insecure because it does not enforce the uniqueness of the sorted elements
+/// To see why this leaks information: assume that all elements are equal,
+/// then quicksort runs in time O(n^2) while for unique elements, it is only expected to run in time O(n log n)
+///
+/// The leakage can be fixed by appending a counter on each element that is unique to the element
+/// This adds another >= log N bits, where N is the amount of elements
+///
+/// This implementation of quicksort is in place and uses a stack instead of recursion
+/// It terminates once the stack is empty
+/// # Errors
+/// Will propagate errors from transport and a few typecasts
+pub async fn quicksort_by_key_insecure<C, K, F, S>(
+    ctx: C,
+    list: &mut [S],
+    desc: bool,
+    f: F,
+) -> Result<(), Error>
+where
+    C: Context,
+    S: Send + Sync,
+    F: Fn(&S) -> &AdditiveShare<K> + Send + Sync,
+    for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
+    K: WeakSharedValue + Field + CustomArray<Element = Boolean>,
+{
+    // create stack
+    let mut stack: Vec<(C, usize, usize)> = vec![];
+
+    // initialize stack
+    stack.push((ctx, 0usize, list.len()));
+
+    // iterate through quicksort recursions
+    while let Some((ctx, b_l, b_r)) = stack.pop() {
+        // start of quicksort function
+        // check whether sort is needed
+        if b_l + 1 < b_r {
+            // set up iterator
+            let mut iterator = list[b_l..b_r].iter();
+            // first element is pivot, apply key extraction function f
+            let pivot = f(iterator.next().unwrap());
+            // create pointer to context for moving into closure
+            let pctx = &ctx;
+            let pf = &f;
+            // precompute comparison against pivot and reveal result in parallel
+            let comp = seq_join(
+                ctx.active_work(),
+                stream_iter(iterator.enumerate().map(|(n, x)| {
+                    async move {
+                        // compare current element against pivot
+                        let sh_comp = compare_gt(
+                            pctx.narrow(&Step::Compare)
+                                .set_total_records(b_r - (b_l + 1)),
+                            RecordId::from(n),
+                            // apply key extraction function f to element x
+                            pf(x),
+                            pivot,
+                        )
+                        .await?;
+
+                        // reveal outcome of comparison
+                        sh_comp
+                            .reveal(
+                                pctx.narrow(&Step::Reveal)
```

Comment:
I think it'd be better to move these calls to `set_total_records` up, outside of the iterator. They only need to be set once per recursion. Subsequent calls to `narrow` will inherit the number of records.